### PR TITLE
chore: optimize resource_class usage

### LINKED ISSUE #16252: Optimize CI resource_class usage
### Current behavior

We use `medium` resource class for all jobs.

### Desired behavior

Use the cheapest possible resource class for all jobs without degrading performance to developers.

### PR DESCRIPTION
**Pull request by flotwig**
Closes #16252

Based on the data collected here: https://docs.google.com/spreadsheets/d/1nkpeZ2dGAq-Es1fKCALiMgWSHCzl6I1PU7y5adSlHes/edit#gid=0

* The default `resource_class` was set to `small` (from `medium`) for all Linux jobs.
* Jobs that increased in runtime by 20 seconds or more when running with `small` (or that OOMed) were bumped to `medium`.
* All bottleneck jobs (`node-modules-install`, `build`, `create-build-artifacts`, `npm-release`) were bumped to `medium+`.
* In the end, 24/54 non-bottleneck jobs were able to run without degradation on `small` (44.444%)
* See spreadsheet column Q+R for job-by-job diff of the changes in this PR

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -1815,7 +1838,7 @@ jobs:
 
   "test-binary-against-cypress-realworld-app":
     <<: *defaults
-    executor: cy-doc-plus
+    resource_class: medium
```
Comment (flotwig) [inline circle.yml:1841]:
```suggestion
    resource_class: medium+
```

**Review by tgriesser — APPROVED** [review]
Great work, excellent analysis here - If we start to see any OOM issues in the future we can always continue to fine tune this further

**Comment by brian-mann** [comment]
The `percy-finalize` job should be switched to a `small` since all it does is poll for results. It doesn't do anything CPU intensive - it'll "hang" open for as long as the jobs its awaiting are running.

**Comment by flotwig** [comment]
https://github.com/cypress-io/cypress/pull/16587

**Pull request merged**